rs/consensus/vetkd/src/lib.rs

```diff
@@ -204,78 +210,91 @@ impl VetKdPayloadBuilderImpl {
             )
         };
 
-        // Iterate over all outstanding VetKD requests
-        let mut candidates = BTreeMap::new();
-        let mut accumulated_size = 0;
-        for (callback_id, context) in state.signature_request_contexts() {
-            if !context.is_vetkd() {
-                // Skip non-vetkd contexts.
-                continue;
-            }
+        let accumulated_size_estimate = AtomicUsize::new(0);
+        let candidates = self.thread_pool.install(|| {
+            state
+                .signature_request_contexts()
+                .par_iter()
+                .flat_map(|(callback_id, context)| {
+                    if !context.is_vetkd() {
+                        // Skip non-vetkd contexts.
+                        return None;
+                    }
 
-            if delivered_ids.contains(callback_id) {
-                // Skip contexts for which we already delivered a response.
-                continue;
-            }
+                    if delivered_ids.contains(callback_id) {
+                        // Skip contexts for which we already delivered a response.
+                        return None;
+                    }
 
-            let candidate = if let Some(reject) =
-                reject_if_invalid(&valid_keys, context, &request_expiry, Some(&self.metrics))
-            {
-                reject
-            } else {
-                let Some(shares) = grouped_shares.get(callback_id) else {
-                    continue;
-                };
-                let ThresholdArguments::VetKd(ctxt_args) = &context.args else {
-                    continue;
-                };
-                let args = VetKdArgs {
-                    context: VetKdDerivationContext {
-                        caller: context.request.sender.into(),
-                        context: context.derivation_path.iter().flatten().cloned().collect(),
-                    },
-                    ni_dkg_id: ctxt_args.ni_dkg_id.clone(),
-                    input: ctxt_args.input.to_vec(),
-                    transport_public_key: ctxt_args.transport_public_key.clone(),
-                };
-                let key_id = context.key_id();
-                match self.crypto.combine_encrypted_key_shares(shares, &args) {
-                    Ok(key) => {
-                        self.metrics
-                            .payload_metrics_inc("vetkd_agreement_completed", &key_id);
-                        let result = VetKdDeriveKeyResult {
-                            encrypted_key: key.encrypted_key,
+                    if let Some(reject) = reject_if_invalid(
+                        &valid_keys,
+                        context,
+                        &request_expiry,
+                        Some(&self.metrics),
+                    ) {
+                        Some((*callback_id, reject))
+                    } else {
```

Comment:
```suggestion
                    if let Some(reject) = reject_if_invalid(
                        &valid_keys,
                        context,
                        &request_expiry,
                        Some(&self.metrics),
                    ) {
                        return Some((*callback_id, reject))
}
```
To avoid the indentation from `else`

Edit: the suggestion is broken, but I think you get my point